stdlib/asm/crypto/fri.masm

```diff
@@ -0,0 +1,613 @@
+#! s.t. next 3 decreasing ( by 1 ) memory addresses hold remaining 6 coefficients of p. Why this descending order of
+#! memory addressing ? - Read point 2 of https://github.com/0xPolygonMiden/miden-vm/issues/592#issue-1481983459
+proc.compute_alpha
+    movup.4
+    dup
+    sub.1
+    movdn.5
+    push.0.0.0.0
+    movup.4
+    mem_loadw
+
+    dup.7
+    dup.7
+    ext2mul
+    ext2add
+
+    movup.6
+    dup
+    sub.1
+    movdn.7
+    push.0.0.0.0
+    movup.4
+    mem_loadw
+
+    movup.5
+    movup.5
+    dup.9
+    dup.9
+    ext2mul
+    ext2add
+
+    dup.7
+    dup.7
+    ext2mul
+    ext2add
+
+    movup.6
+    dup
+    sub.1
+    movdn.7
+    push.0.0.0.0
+    movup.4
+    mem_loadw
+
+    movup.5
+    movup.5
+    dup.9
+    dup.9
+    ext2mul
+    ext2add
+
+    dup.7
+    dup.7
+    ext2mul
+    ext2add
+
+    movup.6
+    push.0.0.0.0
+    movup.4
+    mem_loadw
+
+    movup.5
+    movup.5
+    dup.9
+    dup.9
+    ext2mul
+    ext2add
+
+    movup.7
+    movup.7
+    ext2mul
+    ext2add
+end
+
+#! Given remainder codeword with 64 evaluations ( as starting memory address on stack ), this routine attempts to
+#! probabilistically verify remainder following the technique described in https://github.com/0xPolygonMiden/miden-vm/issues/592
+#! and an optimization of how to efficiently compute β in https://github.com/0xPolygonMiden/miden-vm/pull/644#pullrequestreview-1265141094
+#!
+#! Expected stack state
+#!
+#! [mem_ptr_q, ...]
+#!
+#! A few assumptions about mem_ptr_q
+#!
+#! - mem_ptr_q should be absolute memory address of the beginning of remainder codeword
+#! - Each evaluation is 2 field elements wide because they belong to quadratic extension field over F_p | p = 2^64 - 2^32 + 1
+#! - Meaning each memory address will hold two consecutive evaluations
+#! - Words ( four field elements ), in memory, are laid out in this order (a0_0, a0_1, a1_0, a1_1)
+#! - That means (a0_1, a0_0) -> first evaluation and (a1_1, a1_0) -> next evaluation
+#! - Next 31 memory addresses should be holding remaining 62 evaluations
+#! - If mem_ptr_q holds (a0_0, a0_1, a1_0, a1_1), then mem_ptr_q + 1, must hold (a2_0, a2_1, a3_0, a3_1)
+#! - That way, mem_ptr_q + 31 should be holding (a62_0, a62_1, a63_0, a63_1)
+#!
+#! Final stack looks like
+#!
+#! [...]
+#!
+#! If remainder verification fails, execution of the program stops.
+export.verify_remainder_64.4
+    # [mem_ptr_q, ...]
+    locaddr.0
+    swap
+
+    # [mem_ptr_q, mem_ptr_p, ...]
+    #
+    # note, mem_ptr_p is pointing to function local memory address space
+    # which will be filled up with polynomial p's coefficients, which is computed
+    # outside of the VM, from given codeword q.
+    exec.compute_tau
+
+    # [τ1, τ0, mem_ptr_q, ...]
+    exec.compute_beta
+
+    # [β1, β0, τ1, τ0, mem_ptr_q, ...]
+    movup.4
+    drop
+    locaddr.3
+    movdn.4
+
+    # [β1, β0, τ1, τ0, mem_ptr_p, ...]
```

Comment:
Seems like this block replaces `mem_ptr_q` with `mem_ptr_p`. One thing which is not clear to me is why `mem_ptr_p` is at `locaddr.3`. Wasn't it defined as `locaaddr.0` on line 581 above?

Reply:
Very nice catch ! 

Actually for computing `α` - which is done using Horner's rule, it's easier if polynomial coefficients are stored in descending order in memory, so we access coefficients of `p` in opposite order. 

Read https://github.com/0xPolygonMiden/miden-vm/blob/f0b1d17672364d01a1d69e210613ee5a85d2a8cb/stdlib/asm/crypto/fri.masm#L480-L483 

This Rust code by @Al-Kindi-0 can be easier to read https://github.com/Al-Kindi-0/miden/blob/1b9801c6c0fcaf17047f0e96dd5a46252eaa76a3/miden/tests/integration/stdlib/crypto/fri/verifier_fri_e2f4.rs#L338-L345

Reply:
Makes sense! Could we add a small comment on line 595 to explain this?

Reply:
Sure, done in b433af05